tools/autograd/gen_variable_type.py

```diff
@@ -733,9 +738,52 @@ def gen_variable_type(
         },
     )
 
+    # helper that generates a TORCH_LIBRARY_IMPL macro for each
+    # dispatch key that appears in derivatives.yaml
+    def wrapper_registrations(used_keys: Set[str]) -> str:
+        library_impl_macro_list: List[str] = []
+        for key in used_keys:
+            dispatch_key = key
+            if key == "Default":
+                dispatch_key = "Autograd"
+            library_impl_macro = (
+                f"TORCH_LIBRARY_IMPL(aten, {dispatch_key}, m) "
+                + "{\n"
+                + "${"
+                + f"wrapper_registrations_{key}"
+                + "}\n}"
+            )
+            library_impl_macro_list += [library_impl_macro]
+        return "\n\n".join(library_impl_macro_list)
+
+    # Generate a new template from VariableType.cpp which replaces ${wrapper_registrations}
+    # with per key TORCH_LIBRARY_IMPL macros for each key that appears in derivatives.yaml
```

Comment:
I'm wondering if there isn't a more straightforward way to do this, like a new macro that does the necessary expansion directly rather than generating a new template that's then re-expanded. But I might be overlooking some difficulty. (maybe an example of the generated template would help clarify, if it's easy to throw in a paste)

Reply:
Generated template is pasted in the dropdown below! 
There will be a separate 
```
TORCH_LIBRARY_IMPL(aten, {key}, m) {
${wrapper_registrations_{key}}
}
``` 
for each key that appears in derivatives.yaml! 
<details>
<summary>Here's the generated template</summary>
<br>

```
#include "torch/csrc/autograd/VariableTypeUtils.h"
#include "torch/csrc/autograd/generated/VariableType.h"
#include "torch/csrc/autograd/FunctionsManual.h"

#include <ATen/RedispatchFunctions.h>
#include <ATen/core/TorchDispatchModeTLS.h>
#include <torch/library.h>


// @generated by torchgen/gen.py from VariableType.cpp

// NOTE [Sharded File]: on this file's split-into-shards state
//
// Back in the good old days, VariableType.cpp was generated as one
// file with every function in it, and everything was great and
// simple.
//
// However, this file was also very large (over 36,000 lines), and
// compiling it was very slow, and in fact was a significant
// bottleneck for incremental rebuilds. To address this, we now
// generate the file split across multiple shards, named
// VariableType_0.cpp and so on, which can be compiled in parallel.
//
// For ease of inspection and debugging, so that it's not necessary to
// go rooting around in multiple files, we also generate all the
// functions together in VariableTypeEverything.cpp. This generated
// file is only for convenience; it's not actually used in the
// build. If the file you're looking at now is one of the shards, you
// may want to switch over to the Everything variant to make you
// grepping smoother.

using namespace at;
using namespace torch::autograd::generated;
using namespace torch::autograd::generated::details;


namespace torch { namespace autograd {

namespace VariableType {
namespace{
  C10_UNUSED void reset_grad_accumulator(Variable & self) {
    AutogradMeta* meta = torch::autograd::impl::get_autograd_meta(self);
    if (meta != nullptr) {
      meta->grad_accumulator_.reset();
    }
  }
}

namespace {


${type_derived_method_definitions_AutogradNestedTensor}

${type_derived_method_definitions_AutogradCUDA}

${type_derived_method_definitions_Default}
}
}

namespace {

TORCH_LIBRARY_IMPL(aten, AutogradNestedTensor, m) {
${wrapper_registrations_AutogradNestedTensor}
}

TORCH_LIBRARY_IMPL(aten, AutogradCUDA, m) {
${wrapper_registrations_AutogradCUDA}
}

TORCH_LIBRARY_IMPL(aten, Autograd, m) {
${wrapper_registrations_Default}
}

}

}} // namespace torch::autograd
```
</details>

Reply:
Perhaps a variadic macro would work since there would be a variable number of arguments. However I'm not quite sure how to access individual arguments of __VA_ARGS__ (e.g. the Dispatch macros have `<macro>AND`, `<macro>AND2` etc.) which leads me to believe that unpacking a variable number of arguments in a single macro might be tricky (but I am not familiar with this at all so any thoughts would be appreciated!)

Reply:
I'm happy with the two step codegen here on my end.
Maybe we can leave a comment in the original template to mention this to avoid any confusion?

@bhosmer did you had specific concern about this practice?

Reply:
Just complexity - the codegen is already a very complex build-time process, and introducing the notion of chained codegen seems like a worrisome door to open, if there's a reasonable way to avoid it.

I haven't had bandwidth to think about what a solution might look like (but thanks a bunch Mikayla for the quick turnaround on a sample). But ofc if a solution is worse than the problem - or blocks Mikayla landing this - then I wouldn't insist. Gimme till like EOD to either suggest an alternative or give up 😛

Reply:
@mikaylagawarecki so the straightforward alternative is
* make `${type_derived_method_definitions}` iterate over all used keys and put the definitions out in one big slab
* make `${wrapper_registrations}` put out a `TORCH_LIBRARY_IMPL` block for each used key (and remove the literal `TORCH_LIBRARY_IMPL(...) {...}` text out of the template)

...but tbh it probably isn't worth the trouble - in particular the latter would need some nontrivial new plumbing.

Your current approach is a clever way to avoid all that, which is of course a plus from the POV of code reuse, churn etc. My worry was that having introduced the chained-codegen technique it might get used in less defensible situations, but that's on us to prevent down the line. Definitely not worth the extra work on this PR to avoid.